### ISSUE TITLE
你好，询问一下Multi-band-WaveRNN对于多频带的输入输出

### ISSUE THREAD

**Issue by Rongjiehuang** [yanggeng1995/subband_WaveRNN#1]
你好！想请教一下！Multi-band-WaveRNN对于多频带的输入输出
我使用的是fatchord_version.的wavernn，代码放在我的仓库中。但是生成不出有意义的声音，我怀疑模型输入的时候mel condition出了问题。想请教一下您：
我是将：四个频带音频x维度为(batch,T,4)，mel为(batch,T,80)，上采样得到的残差边为(batch,T,32)。将他们cat后传入网络中，因此网络不同之处是输入音频x由(batch,T,1)变为(batch,T,4)，同时输出使用四个全连接层生成四个频带音频，并且得到四个频带的loss相加并反向传播。
我查看了论文DURIAN中图，输入特征四个频带的音频似乎是也是在高维上(batch,T,4)扩展而不是时间步上扩展成(batch,4T,1)，想向您证实并交流一下，谢谢！

训练部分：

       for i, (x, y, m) in enumerate(train_set, 1):
            x, m, y = x.to(device), m.to(device), y.to(device)  # x/y: (Batch, sub_bands, T)

#########################  MultiBand-WaveRNN   #########################
            if hp.voc_multiband:
                y0 = y[:, 0, :].squeeze(0).unsqueeze(-1)  # y0/y1/y2/y3: (Batch, T, 1)
                y1 = y[:, 1, :].squeeze(0).unsqueeze(-1)
                y2 = y[:, 2, :].squeeze(0).unsqueeze(-1)
                y3 = y[:, 3, :].squeeze(0).unsqueeze(-1)

                y_hat = model(x, m)  # (Batch, T, num_classes, sub_bands)

                if model.mode == 'RAW':
                    y_hat0 = y_hat[:, :, :, 0].transpose(1,2).unsqueeze(-1)  # (Batch, num_classes, T, 1)
                    y_hat1 = y_hat[:, :, :, 1].transpose(1,2).unsqueeze(-1)
                    y_hat2 = y_hat[:, :, :, 2].transpose(1,2).unsqueeze(-1)
                    y_hat3 = y_hat[:, :, :, 3].transpose(1,2).unsqueeze(-1)

                elif model.mode == 'MOL':
                    y0 = y0.float()
                    y1 = y1.float()
                    y2 = y2.float()
                    y3 = y3.float()

                loss = loss_func(y_hat0, y0) + loss_func(y_hat1, y1) + loss_func(y_hat2, y2) + loss_func(y_hat3, y3)

模型结构：

    def forward(self, x, mels):  # x: (Batch, Subband, T)
        device = next(self.parameters()).device  # use same device as parameters

        # Although we `_flatten_parameters()` on init, when using DataParallel
        # the model gets replicated, making it no longer guaranteed that the
        # weights are contiguous in GPU memory. Hence, we must call it again
        self._flatten_parameters()

        self.step += 1
        bsize = x.size(0)
        h1 = torch.zeros(1, bsize, self.rnn_dims, device=device)
        h2 = torch.zeros(1, bsize, self.rnn_dims, device=device)
        mels, aux = self.upsample(mels)

        aux_idx = [self.aux_dims * i for i in range(5)]
        a1 = aux[:, :, aux_idx[0]:aux_idx[1]]
        a2 = aux[:, :, aux_idx[1]:aux_idx[2]]
        a3 = aux[:, :, aux_idx[2]:aux_idx[3]]
        a4 = aux[:, :, aux_idx[3]:aux_idx[4]]
        # x = torch.cat([x.unsqueeze(-1), mels, a1], dim=2)

        x = torch.cat([x.transpose(1,2), mels, a1], dim=2)  # (batch,T,4)  (batch,T,80) (batch,T,32)
        x = self.I(x)  # (batch, T, 116) -> # (batch, T, 512)
        res = x
        # x, _ = self.rnn1(x, h1)
        x, _ = self.rnn1(x)   # 不加入隐藏层-Begee  # (batch, T, 512) -> (batch, T, 512)
        x = x + res    # (batch, T, 512)
        res = x
        x = torch.cat([x, a2], dim=2)  # (batch, T, 512) -> (batch, T, 512+128)
        # x, _ = self.rnn2(x, h2)
        x, _ = self.rnn2(x)  # 不加入隐藏层-Begee  (batch, T, 512+128) -> (batch, T, 512)

        x = x + res
        x = torch.cat([x, a3], dim=2) # (batch, T, 512+128)
        x = F.relu(self.fc1(x)) # (batch, T, 512+128) -> (batch, T, 512)

        x = torch.cat([x, a4], dim=2)  # (batch, T, 512+128)
        x = F.relu(self.fc2(x))  # (batch, T, 512+128) -> (batch, T, 512)

        out0 = self.fc30(x).unsqueeze(-1) # (batch, T, 512) -> (batch, T, 512)
        out1 = self.fc31(x).unsqueeze(-1)
        out2 = self.fc32(x).unsqueeze(-1)
        out3 = self.fc33(x).unsqueeze(-1)
        out = torch.cat([out0,out1,out2,out3], dim=3)  # (B, T, num_classes, sub_band)
        return out

感谢您的解答！

**Comment by yanggeng1995** [comment]
这个模型的框架可能不是很好训练，你用的mol loss的话，可能子带需要调一些参数什么的，类似于残差那种结构其实没必要用在网络里面。
关于子带部分，输入是condition[B, T, lc_dim]和多个子带音频[B, T,
4]，模型最终的输入你可以直接用一层dnn直接预测四维，我当时在git上写的用四个dnn是
考虑到后面方便做工程优化进行同时预测，但是好像那样会不太好。关于全频带WaveRNN你可以参考我在git上实现的（
https://github.com/yanggeng1995/WaveRNN ），感觉会比这个repo效果好点。
我私下的子带WaveRNN就是基于这个版本做的。

Rongjiehuang <notifications@github.com> 于2021年1月4日周一 下午10:53写道：

**Comment by Rongjiehuang** [comment]
您好！感谢解答。deep_mind version确实把每一训练步作为ceil，每个样本每轮训练会多次进入forward，而fatchord_version并不是ceil，每个样本每轮训练只进入一次forward。因而如果只改变最后一层全连接层，最后一层分出生成四个子带，最后全连接层能力不足，生成的子带效果差。
因此需要增加网络学习四个子带差异性的能力，我目前直接从rnn开始不共享参数，这样学习到的子带以及很不错了，但是训练速度慢，multiband的快速训练优势消失了。后面想是用一些好一点的办法，请问您是怎么看的呢

**Comment by yanggeng1995** [comment]
You can leave an email and I will reply you by email.

**Comment by Rongjiehuang** [comment]
thank you, my email is rongjiehuang@zju.edu.cn
